staking/programs/staking/src/state/positions.rs

```diff
@@ -827,4 +989,414 @@ pub mod tests {
+            // epoch
+            if (target
+                == &TargetWithParameters::IntegrityPool {
+                    publisher: FIRST_PUBLISHER,
+                })
+                && prev_state != &PositionState::LOCKED
+                && prev_state != &PositionState::PREUNLOCKING
+                && pre_position_buckets.get(&(*target, *prev_state, *curr_state))
+                    != post_position_buckets.get(&(*target, *prev_state, *curr_state))
+            {
+                return false;
+            }
+
+            // slashing affects positions that are locked or preunlocking with ratio slash_ratio
+            if (target
+                == &TargetWithParameters::IntegrityPool {
+                    publisher: FIRST_PUBLISHER,
+                })
+                && (prev_state == &PositionState::LOCKED
+                    || prev_state == &PositionState::PREUNLOCKING)
+            {
+                let expected_slashed: u64 = (*pre_position_buckets
+                    .get(&(*target, *prev_state, *curr_state))
+                    .unwrap() as u128
+                    * slash_ratio as u128
+                    / 1_000_000)
+                    .try_into()
+                    .unwrap();
+                let pre_slashable = *pre_position_buckets
+                    .get(&(*target, *prev_state, *curr_state))
+                    .unwrap();
+                let post_slashable = *post_position_buckets
+                    .get(&(*target, *prev_state, *curr_state))
+                    .unwrap_or(&0);
+                let slashed = pre_slashable - post_slashable;
+
+                if !((expected_slashed >= slashed)
+                    && (slashed >= expected_slashed.saturating_sub(next_index.into())))
+                {
+                    return false;
+                }
+
+                if prev_state == &PositionState::LOCKED {
+                    post_amount_slashable_locked += post_slashable;
+                }
+                if prev_state == &PositionState::PREUNLOCKING
+                    && curr_state == &PositionState::PREUNLOCKING
+                {
+                    post_amount_slashed_preunlocking += post_slashable;
+                }
+            }
+
+
+            // slashing reduces governance positions
+            if (target == &TargetWithParameters::Voting)
+                && pre_position_buckets.get(&(*target, *prev_state, *curr_state))
+                    < post_position_buckets.get(&(*target, *prev_state, *curr_state))
+            {
+                return false;
+            }
+        }
+
+        // the returned values match the position updates
+        if locked_slashed != amount_slashable_locked - post_amount_slashable_locked {
+            return false;
+        }
+
+        if preunlocking_slashed != amount_slashable_preunlocking - post_amount_slashed_preunlocking
+        {
+            return false;
```

Comment:
does returning false mean that the test will fail?

Reply:
Yes thought it's very hard to find where the false was returned.
I think this is useful because when quicktest sees that a tests fails with a given input it tries to rerun it with smaller inputs derived from the input so that hopefully it can produce a minimal input that fails